### ISSUE TITLE
LDAP error: operations error!

### ISSUE THREAD

**Issue by marek-cerny** [ibv/LDAP-Admin#3]
Hello.

After successful compilation, I am encountering this error when attempting to connect to your AD controller (running on Samba):
![ldapadmin-error](https://user-images.githubusercontent.com/24892664/31265756-bc7febce-aa70-11e7-938c-9f4e59ad1740.jpg)

Lazarus IDE gives some more info about the error:
![ldapadmin-debug](https://user-images.githubusercontent.com/24892664/31265768-c726f4e6-aa70-11e7-962c-98529a79e6c9.jpg)

Connection properties:
![ldapadmin-config](https://user-images.githubusercontent.com/24892664/31265780-d95d1b18-aa70-11e7-98a5-2fac819fadb0.jpg)

When using the official Windows version of LDAP Admin, the same connection works flawlessly.

Compiled on Debian 9.1, Lazarus IDE 1.6.2.

Any idea what might be wrong?

Thanks!

**Comment by ibv** [comment]
I tried to enter incorrect login data in Win AD and the same error occurred,
are your name and password really correct?

**Comment by rabid-user** [comment]
Maybe a bit late for this answer, but I had this same error when attempting to use ldapadmin for an SSL connection, running on a linux mint 18.2 release. Since I knew the user/password were correct, I expected it must be an issue with SSL.

I ran the debugger in lazarus and found that the libraries libcrypto.so and libssl.so were not being loaded even though they exist on my system.

I needed to run the following commands, and then it worked.
sudo ln -s libcrypto.so.1.0.0 /lib/x86_64-linux-gnu/libcrypto.so
sudo ln -s libssl.so.1.0.0 /lib/x86_64-linux-gnu/libssl.so
sudo ldconfig -n -v /lib/x86_64-linux-gnu

**Comment by bmaupin** [comment]
Same problem on Ubuntu 18.04. I think I figured it out.

The Synapse SSL library used by LDAP-Admin [looks for "libssl.so" and "libcrypto.so"](https://github.com/ibv/LDAP-Admin/blob/master/Source/synapse/ssl_openssl_lib.pas#L106:L129). On my system, these link to libssl.so.1.1 and libcrypto.so.1.1:

```
$ ldconfig -p | grep libssl.so
	...
	libssl.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libssl.so
$ ls -l /usr/lib/x86_64-linux-gnu/libssl.so
lrwxrwxrwx 1 root root 13 Dec  5 10:59 /usr/lib/x86_64-linux-gnu/libssl.so -> libssl.so.1.1
$ ldconfig -p | grep libcrypto.so
	...
	libcrypto.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libcrypto.so
$ ls -l /usr/lib/x86_64-linux-gnu/libcrypto.so
lrwxrwxrwx 1 root root 16 Dec  5 10:59 /usr/lib/x86_64-linux-gnu/libcrypto.so -> libcrypto.so.1.1
```

I can confirm this by running this command while LDAP-Admin is open:

```
$ lsof -p $(pidof LdapAdmin) | grep mem | egrep -i "ssl|crypt"
LdapAdmin 25940 user  mem       REG               8,17   433760  1182086 /usr/lib/x86_64-linux-gnu/libssl.so.1.1
LdapAdmin 25940 user  mem       REG               8,17  2575848  1182023 /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
```

My best guess is that the Synaptic SSL libraries aren't compatible with SSL/crypto 1.1. If I change all instances of `libssl.so` to `libssl.so.1.0.0` and `libcrypto.so` to `libcrypto.so.1.0.0` and then rebuild, it works. And I can confirm those libaries are being used:

```
$ lsof -p $(pidof LdapAdmin) | grep mem | egrep -i "ssl|crypt"
LdapAdmin 25397 user  mem       REG               8,17   426232  1181336 /usr/lib/x86_64-linux-gnu/libssl.so.1.0.0
LdapAdmin 25397 user  mem       REG               8,17  2357760  1181335 /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
```

(I'm not sure why the reference in the `OS2` section needs to be changed, but for whatever reason that seems to make a difference. Possibly another bug).

I'm guessing those libraries are installed by default, but if not you can install them by installing the `libssl1.0.0` package in Ubuntu.

> I tried to enter incorrect login data in Win AD and the same error occurred,
are your name and password really correct?

This problem will occur even when connecting anonymously, so it's definitely not a username/password issue.

Hope that helps!

**Comment by bmaupin** [comment]
This issue is fixed for me by ffa15ba.
